File: sdk/src/trading_sdk/trading.py
from typing_extensions import Protocol, TypedDict, Literal, TypeVar, NotRequired
from decimal import Decimal

from trading_sdk.types import Side, TimeInForce, Num
from trading_sdk.errors import AuthedError

S = TypeVar('S', bound=str)

class Trading(Protocol):

  class BaseOrder(TypedDict):
    side: Side

  class BaseLimitOrder(BaseOrder):
    qty: Num
    price: Num
    time_in_force: NotRequired[TimeInForce]

  class LimitOrder(BaseLimitOrder):
    type: Literal['LIMIT']

  class LimitMakerOrder(BaseLimitOrder):
    type: Literal['LIMIT_MAKER']

  class MarketOrder(BaseOrder):
    type: Literal['MARKET']
    qty: Num
    time_in_force: NotRequired[TimeInForce]

  Order = LimitOrder | LimitMakerOrder | MarketOrder

  OrderStatus = Literal['NEW', 'PARTIALLY_FILLED', 'FILLED', 'CANCELED']
  
  class PlaceOrderResponse(TypedDict):
    order_id: str

  async def place_order(self, symbol: str, order: Order) -> PlaceOrderResponse | AuthedError:
    ...

  async def cancel_order(self, symbol: str, *, order_id: str) -> AuthedError | None:
    ...

  async def cancel_all_orders(self, symbol: str) -> AuthedError | None:
    ...

  class QueryOrderResponse(TypedDict):
    status: 'Trading.OrderStatus'
    executed_qty: Decimal

  async def query_order(self, symbol: str, *, order_id: str) -> QueryOrderResponse:
    ...

  class Balance(TypedDict):
    free: Decimal
    locked: Decimal

  async def get_balances(self, *currencies: S) -> dict[S, Balance]:
    ...

File: rust/src/trading.rs
use crate::types::{Side, TimeInForce};
use crate::errors::AuthedError;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LimitOrder {
  pub side: Side,
  pub qty: String,
  pub price: String,
  pub time_in_force: Option<TimeInForce>,
  pub post_only: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MarketOrder {
  pub side: Side,
  pub qty: String,
  pub time_in_force: Option<TimeInForce>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Order {
  Limit(LimitOrder),
  Market(MarketOrder),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PlaceOrderResponse {
  pub order_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OrderStatus {
  New,
  PartiallyFilled,
  Filled,
  Canceled,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QueryOrderResponse {
  pub status: OrderStatus,
  pub executed_qty: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Balance {
  pub free: String,
  pub locked: String,
}

#[async_trait::async_trait]
pub trait Trading {
  async fn place_order(&self, symbol: &str, order: Order) -> Result<PlaceOrderResponse, AuthedError>;
  async fn cancel_order(&self, symbol: &str, order_id: &str) -> Result<(), AuthedError>;
  async fn cancel_all_orders(&self, symbol: &str) -> Result<(), AuthedError>;
  async fn query_order(&self, symbol: &str, order_id: &str) -> Result<QueryOrderResponse, AuthedError>;
  async fn get_balance(&self, currency: &str) -> Result<Balance, AuthedError>;
} 
